flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -1368,13 +1369,19 @@ private void checkpointStreamOperator(StreamOperator<?> op) throws Exception {
 
 	private class TimerInvocationContext implements SystemProcessingTimeService.ScheduledCallbackExecutionContext {
 		@Override
-		public void invoke(ProcessingTimeCallback callback, long timestamp) {
-			synchronized (getCheckpointLock()) {
-				try {
-					callback.onProcessingTime(timestamp);
-				} catch (Throwable t) {
-					handleAsyncException("Caught exception while processing timer.", new TimerException(t));
-				}
+		public void invoke(ProcessingTimeCallback callback, long timestamp) throws InterruptedException {
+			try {
+				mailboxProcessor.getMailboxExecutor(TaskMailbox.MAX_PRIORITY).execute(() -> {
+					synchronized (getCheckpointLock()) {
+						try {
+							callback.onProcessingTime(timestamp);
+						} catch (Throwable t) {
+							handleAsyncException("Caught exception while processing timer.", new TimerException(t));
+						}
+					}
+				});
+			} catch (Throwable t) {
```

Comment:
Do we need the outer exception handling? If the mailbox is incorrect state, then I'd image that the stack trace is already good enough.